File: dart/lib/list_converters.dart
import 'deva_gujr_converter.dart';
import 'transliteration_result.dart';

/// Extension methods for performing bulk / list transliteration on an [Iterable] of [String] items.
extension IndicScriptListConversion on Iterable<String> {
  /// Translates each IAST string in the collection to Devanagari returning a [List] of [TransliterationResult] objects.
  List<TransliterationResult> toDevanagari({
    IastToDevanagariOptions options = const IastToDevanagariOptions(),
  }) =>
      map((s) => s.toDevanagari(options: options)).toList();

  /// Translates each IAST string in the collection to Gujarati returning a [List] of [TransliterationResult] objects.
  List<TransliterationResult> toGujarati({
    IastToGujaratiOptions options = const IastToGujaratiOptions(),
  }) =>
      map((s) => s.toGujarati(options: options)).toList();

  /// Translates each IAST string in the collection to Plain English returning a [List] of [TransliterationResult] objects.
  List<TransliterationResult> toPlainEnglish({
    IastPlainEnglishOptions options = const IastPlainEnglishOptions(),
  }) =>
      map((s) => s.toPlainEnglish(options: options)).toList();

  /// Converts each IAST string in the collection to Devanagari script strings.
  List<String> toDevanagariFromIast({
    IastToDevanagariOptions options = const IastToDevanagariOptions(),
  }) =>
      map((s) => s.toDevanagariFromIast(options: options)).toList();

  /// Converts each IAST string in the collection to Gujarati script strings.
  List<String> toGujaratiFromIast({
    IastToGujaratiOptions options = const IastToGujaratiOptions(),
  }) =>
      map((s) => s.toGujaratiFromIast(options: options)).toList();

  /// Converts each IAST string in the collection to Plain English strings.
  List<String> toPlainEnglishFromIast({
    IastPlainEnglishOptions options = const IastPlainEnglishOptions(),
  }) =>
      map((s) => s.toPlainEnglishFromIast(options: options)).toList();

  /// Converts each Devanagari string in the collection back to canonical IAST.
  List<String> toCanonicalIastFromDevanagari({
    ScriptToIastOptions options = const ScriptToIastOptions(),
  }) =>
      map((s) => s.toCanonicalIastFromDevanagari(options: options)).toList();

  /// Converts each Gujarati string in the collection back to canonical IAST.
  List<String> toCanonicalIastFromGujarati({
    ScriptToIastOptions options = const ScriptToIastOptions(),
  }) =>
      map((s) => s.toCanonicalIastFromGujarati(options: options)).toList();

  /// Converts each Devanagari string in the collection to canonical Gujarati.
  List<String> toCanonicalGujaratiFromDevanagari({
    IndicScriptConversionOptions options = const IndicScriptConversionOptions(),
  }) =>
      map((s) => s.toCanonicalGujaratiFromDevanagari(options: options))
          .toList();

  /// Converts each Gujarati string in the collection to canonical Devanagari.
  List<String> toCanonicalDevanagariFromGujarati({
    IndicScriptConversionOptions options = const IndicScriptConversionOptions(),
  }) =>
      map((s) => s.toCanonicalDevanagariFromGujarati(options: options))
          .toList();

  /// Converts each Devanagari string in the collection to Gujarati with exact source metadata recovery.
  List<String> toGujaratiFromDevanagari({
    IndicScriptConversionOptions options = const IndicScriptConversionOptions(),
  }) =>
      map((s) => s.toGujaratiFromDevanagari(options: options)).toList();

  /// Converts each Gujarati string in the collection to Devanagari with exact source metadata recovery.
  List<String> toDevanagariFromGujarati({
    IndicScriptConversionOptions options = const IndicScriptConversionOptions(),
  }) =>
      map((s) => s.toDevanagariFromGujarati(options: options)).toList();

  /// Recovers exact original Devanagari from each Gujarati string in the collection.
  List<String> toExactDevanagariFromGujarati() =>
      map((s) => s.toExactDevanagariFromGujarati()).toList();

  /// Recovers exact original Gujarati from each Devanagari string in the collection.
  List<String> toExactGujaratiFromDevanagari() =>
      map((s) => s.toExactGujaratiFromDevanagari()).toList();

  /// Recovers exact original IAST from each Devanagari string in the collection.
  List<String> toExactIastFromDevanagari() =>
      map((s) => s.toExactIastFromDevanagari()).toList();

  /// Recovers exact original IAST from each Gujarati string in the collection.
  List<String> toExactIastFromGujarati() =>
      map((s) => s.toExactIastFromGujarati()).toList();
}

/// Converts a list of IAST strings to Devanagari script strings.
List<String> toDevanagariFromIastList(
  Iterable<String> items, {
  IastToDevanagariOptions options = const IastToDevanagariOptions(),
}) =>
    items.toDevanagariFromIast(options: options);

/// Converts a list of IAST strings to Gujarati script strings.
List<String> toGujaratiFromIastList(
  Iterable<String> items, {
  IastToGujaratiOptions options = const IastToGujaratiOptions(),
}) =>
    items.toGujaratiFromIast(options: options);

/// Converts a list of IAST strings to Plain English strings.
List<String> toPlainEnglishFromIastList(
  Iterable<String> items, {
  IastPlainEnglishOptions options = const IastPlainEnglishOptions(),
}) =>
    items.toPlainEnglishFromIast(options: options);

/// Converts a list of Devanagari strings to canonical Gujarati script strings.
List<String> toCanonicalGujaratiFromDevanagariList(
  Iterable<String> items, {
  IndicScriptConversionOptions options = const IndicScriptConversionOptions(),
}) =>
    items.toCanonicalGujaratiFromDevanagari(options: options);

/// Converts a list of Gujarati strings to canonical Devanagari script strings.
List<String> toCanonicalDevanagariFromGujaratiList(
  Iterable<String> items, {
  IndicScriptConversionOptions options = const IndicScriptConversionOptions(),
}) =>
    items.toCanonicalDevanagariFromGujarati(options: options);

/// Recovers exact original Devanagari strings from a list of Gujarati strings.
List<String> toExactDevanagariFromGujaratiList(Iterable<String> items) =>
    items.toExactDevanagariFromGujarati();

/// Recovers exact original Gujarati strings from a list of Devanagari strings.
List<String> toExactGujaratiFromDevanagariList(Iterable<String> items) =>
    items.toExactGujaratiFromDevanagari();

File: javascript/src/list-converters.js
import { toDevanagariFromIast } from './latn-iast-to-deva.js';
import { toGujaratiFromIast } from './latn-iast-to-gujr.js';
import { toPlainEnglishFromIast } from './latn-iast-transcription.js';
import {
    toIastFromDevanagari,
    toIastFromGujarati,
    toCanonicalIastFromDevanagari,
    toCanonicalIastFromGujarati,
    toExactIastFromDevanagari,
    toExactIastFromGujarati,
} from './brahmic-to-latn-iast.js';
import {
    toDevanagariFromGujarati,
    toGujaratiFromDevanagari,
    toCanonicalGujaratiFromDevanagari,
    toCanonicalDevanagariFromGujarati,
    toExactDevanagariFromGujarati,
    toExactGujaratiFromDevanagari,
} from './deva-gujr-converter.js';
import { toDevanagari, toGujarati, toPlainEnglish } from './transliteration-result.js';

/**
 * Bulk converts an array of IAST strings to Devanagari script strings.
 * @param {Array<string>} items Array of strings to convert.
 * @param {object} [options] Conversion options.
 * @returns {Array<string>} Array of converted Devanagari strings.
 */
export function toDevanagariFromIastList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toDevanagariFromIast(item, options));
}

/**
 * Bulk converts an array of IAST strings to Gujarati script strings.
 * @param {Array<string>} items Array of strings to convert.
 * @param {object} [options] Conversion options.
 * @returns {Array<string>} Array of converted Gujarati strings.
 */
export function toGujaratiFromIastList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toGujaratiFromIast(item, options));
}

/**
 * Bulk converts an array of IAST strings to Plain English strings.
 * @param {Array<string>} items Array of strings to convert.
 * @param {object} [options] Conversion options.
 * @returns {Array<string>} Array of converted Plain English strings.
 */
export function toPlainEnglishFromIastList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toPlainEnglishFromIast(item, options));
}

/**
 * Smart converts an array of Devanagari strings back to IAST (recovers exact source if embedded metadata exists).
 * @param {Array<string>} items Array of Devanagari strings.
 * @param {object} [options] Conversion options.
 * @returns {Array<string>} Array of IAST strings.
 */
export function toIastFromDevanagariList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toIastFromDevanagari(item, options));
}

/**
 * Smart converts an array of Gujarati strings back to IAST (recovers exact source if embedded metadata exists).
 * @param {Array<string>} items Array of Gujarati strings.
 * @param {object} [options] Conversion options.
 * @returns {Array<string>} Array of IAST strings.
 */
export function toIastFromGujaratiList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toIastFromGujarati(item, options));
}

/**
 * Bulk converts an array of Devanagari strings back to canonical IAST.
 * @param {Array<string>} items Array of Devanagari strings.
 * @param {object} [options] Conversion options.
 * @returns {Array<string>} Array of canonical IAST strings.
 */
export function toCanonicalIastFromDevanagariList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toCanonicalIastFromDevanagari(item, options));
}

/**
 * Bulk converts an array of Gujarati strings back to canonical IAST.
 * @param {Array<string>} items Array of Gujarati strings.
 * @param {object} [options] Conversion options.
 * @returns {Array<string>} Array of canonical IAST strings.
 */
export function toCanonicalIastFromGujaratiList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toCanonicalIastFromGujarati(item, options));
}

/**
 * Bulk recovers exact original IAST strings from an array of Devanagari strings.
 * @param {Array<string>} items Array of Devanagari strings.
 * @returns {Array<string>} Array of exact IAST strings.
 */
export function toExactIastFromDevanagariList(items) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toExactIastFromDevanagari(item));
}

/**
 * Bulk recovers exact original IAST strings from an array of Gujarati strings.
 * @param {Array<string>} items Array of Gujarati strings.
 * @returns {Array<string>} Array of exact IAST strings.
 */
export function toExactIastFromGujaratiList(items) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toExactIastFromGujarati(item));
}

/**
 * Bulk converts an array of Devanagari strings to Gujarati strings (recovers exact source metadata if present).
 * @param {Array<string>} items Array of Devanagari strings.
 * @param {object} [options] Conversion options.
 * @returns {Array<string>} Array of Gujarati strings.
 */
export function toGujaratiFromDevanagariList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toGujaratiFromDevanagari(item, options));
}

/**
 * Bulk converts an array of Gujarati strings to Devanagari strings (recovers exact source metadata if present).
 * @param {Array<string>} items Array of Gujarati strings.
 * @param {object} [options] Conversion options.
 * @returns {Array<string>} Array of Devanagari strings.
 */
export function toDevanagariFromGujaratiList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toDevanagariFromGujarati(item, options));
}

/**
 * Bulk converts an array of Devanagari strings to canonical Gujarati strings.
 * @param {Array<string>} items Array of Devanagari strings.
 * @param {object} [options] Conversion options.
 * @returns {Array<string>} Array of canonical Gujarati strings.
 */
export function toCanonicalGujaratiFromDevanagariList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toCanonicalGujaratiFromDevanagari(item, options));
}

/**
 * Bulk converts an array of Gujarati strings to canonical Devanagari strings.
 * @param {Array<string>} items Array of Gujarati strings.
 * @param {object} [options] Conversion options.
 * @returns {Array<string>} Array of canonical Devanagari strings.
 */
export function toCanonicalDevanagariFromGujaratiList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toCanonicalDevanagariFromGujarati(item, options));
}

/**
 * Bulk recovers exact original Devanagari strings from an array of Gujarati strings.
 * @param {Array<string>} items Array of Gujarati strings.
 * @returns {Array<string>} Array of exact Devanagari strings.
 */
export function toExactDevanagariFromGujaratiList(items) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toExactDevanagariFromGujarati(item));
}

/**
 * Bulk recovers exact original Gujarati strings from an array of Devanagari strings.
 * @param {Array<string>} items Array of Devanagari strings.
 * @returns {Array<string>} Array of exact Gujarati strings.
 */
export function toExactGujaratiFromDevanagariList(items) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toExactGujaratiFromDevanagari(item));
}

/**
 * Bulk converts an array of IAST strings returning an array of Devanagari TransliterationResult envelopes.
 * @param {Array<string>} items Array of strings.
 * @param {object} [options] Conversion options.
 * @returns {Array<TransliterationResult>} Array of result envelopes.
 */
export function toDevanagariList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toDevanagari(item, options));
}

/**
 * Bulk converts an array of IAST strings returning an array of Gujarati TransliterationResult envelopes.
 * @param {Array<string>} items Array of strings.
 * @param {object} [options] Conversion options.
 * @returns {Array<TransliterationResult>} Array of result envelopes.
 */
export function toGujaratiList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toGujarati(item, options));
}

/**
 * Bulk converts an array of IAST strings returning an array of Plain English TransliterationResult envelopes.
 * @param {Array<string>} items Array of strings.
 * @param {object} [options] Conversion options.
 * @returns {Array<TransliterationResult>} Array of result envelopes.
 */
export function toPlainEnglishList(items, options) {
    if (!Array.isArray(items)) throw new TypeError('Expected an array of strings.');
    return items.map((item) => toPlainEnglish(item, options));
}
